Jc/oc 3105/osc authz

jamesc opened the pull request:
auth_info implementation for OSC

franken-hacked into chef_wm_base but will be refactored into mixins later

Commit:
minimal WIP implementation of OSC authz checking

Commit:
more endpoints implemented

Commit:
Flesh out auth_info for all endpoints
Split out lower level authz calls into a separate module
(chef_wm_authz).

Still to do is to reduce duplication in handle_auth_info

Commit:
DRY up the handle_auth_info code

seth commented on src/chef_wm_authz.erl:
```diff
@@ -0,0 +1,48 @@
+-spec is_admin_or_requesting_node(#chef_client{}, binary()) -> boolean().
+is_admin_or_requesting_node(#chef_client{name = Name} = Client, NodeName) ->
+    case NodeName of
+        N when N =:= Name ->
```
You can just match like:

  case NodeName of
      Name -> % they are the same

seth replied:
Although maybe simpler to do that in the fun head:

```
is_admin_or_requesting_node(#chef_client{name = Name}, Name) ->
  true;
is_admin_or_requesting_node(#chef_client{admin = IsAdmin}, _) ->
    true =:= IsAdmin. % or use is_admin, but since it is just the field checking, I don't mind it inline.
```

seth commented on src/chef_wm_base.erl:
```diff
@@ -196,15 +197,19 @@ malformed_request_message(Reason, Req, #base_state{resource_mod=Mod}=State) ->
         {_, Req1, State1} ->
-            {false, Req1, State1}
+            case handle_auth_info(Mod, Req1, State1) of
+                false ->
```
Totally fine with just sending this as-is, but wonder if it would be more clear to have handle_auth_info return 'authorized' | 'no_auth' or 'forbidden' to help avoid the true is false thing.

seth commented:
minor fixes as commented if you want. Otherwise, merge when ready.

Commit:
review fixups
- rename is_FOO authz methods to allow_FOO
- return authorized | forbidden instead of boolean()
- some small matching refactors
